### ISSUE TITLE
Sliding window functions 

### ISSUE THREAD

**Issue by cy6erGn0m** [Kotlin/KEEP#11]
Discussions about the [Sliding window functions](../blob/master/proposals/stdlib/window-sliding.md) will be held here.

**Comment by ilya-g** [comment]
Use cases:

>  fun view(page: Int) = source.slidingWindow(pageSize).drop(page).firstOrNull()

- it might be better to use `elementAtOrNull(n)` instead of `drop(n).firstOrNull()`
- I wouldn't call this an effective implementation for `view(page)`, as it requires to populate and then throw away all pages before the required one. Might be better just `drop(pageSize * page).take(pageSize)`

> permutations generation, advanced signal processing, hex to byte array

- windows returned could have `size < windowSize`, it would break these usages.

**Comment by ilya-g** [comment]
Alternatives:

- consider imperative loops when lazy behavior isn't required.

**Comment by voddan** [comment]
what `slidingWindowIndices` is supposed t do? What is the signature?

**Comment by cy6erGn0m** [comment]
It returns a sequence of ranges rather than sub-contents. Not sure we have to make it public, most likely we shouldn't

**Comment by voddan** [comment]
@cy6erGn0m so it is equivalent to `this.indices.slidingWindow()` ? I don't think this worth a separate function.

**Comment by cy6erGn0m** [comment]
actually it is equivalent to `indices.slidingWindow(..).map { IntRange(it.first(), it.last()) }` that in fact is too inefficient. The only use-case for this function is the case when you don't want to copy sub contents but just want to keep index ranges for performance reasons while `indices.slidingWindow(...)` is definitely not a good idea if you need performance.

**Comment by voddan** [comment]
Yes, now I see the benefit of this function.

What troubles me is that  `slidingWindowIndices` does not depend on the content of the collection. 

The use case you described may be covered with `slidingWindow` overloaded for `ClosedRange<T>` (`IntRange`) with a usage like: `list.indices.slidingWindow()`

**Comment by cy6erGn0m** [comment]
Yes, but it leads to an ambiguity: `IntRange` is both `ClosedRange` and `Iterable<Int>` so having two functions with different return types could cause confusion

**Comment by ilya-g** [comment]
TODO: Elaborate use-cases, remove unnatural ones (paging, bytesFromHex).

**Comment by cy6erGn0m** [comment]
@ilya-g I don't agree that `bytesFromHex` is unnatural: I do like this in tests (with simplified implementation of slidingWindow) as it is one of the easiest ways to do that.

**Comment by ilya-g** [comment]
Regarding `byteFromHex`, I would write it as follows:
```
fun String.hexToBytes(): ByteArray {
    require(this.length % 2 == 0) { "Must contain even number of hex digits." }
    val size = this.length / 2 
    return ByteArray(size) { p -> (Character.digit(this[p*2],16)*16 + Character.digit(this[p*2+1], 16)).toByte() }
}
```
It might look not so clear as with `slidingWindow`, but it's definitely more performant.

Again why it's not the use-case of this proposal — if one need to convert base16 string to a `ByteArray`, he would use some `byteFromHex` function rather than resort to ad-hoc `slidingWindow` every time.

**Comment by cy6erGn0m** [comment]
@ilya-g in your implementation there is too much arithmetic

Of course there are a lot of implementations available, for example, apache commons codec has one for that purpose. The main idea of the example is to show text-processing use-case. Don't get any ideas about text formats with fixed-size fields, don't you?

**Comment by ilya-g** [comment]
> Don't get any ideas about text formats with fixed-size fields, don't you?

No, I don't. It must be crystal clear from the proposal without guessing.

**Comment by ilya-g** [comment]
+ similar API: `StreamEx.ofSubLists` https://github.com/amaembo/streamex/blob/master/src/main/java/one/util/streamex/StreamEx.java#L2639

**Comment by hastebrot** [comment]
Similar API: For Ruby there is not only `Enumerable.each_slice` but also `Enumerable.each_cons`.

**Comment by hastebrot** [comment]
Also similar API: For Clojure there is `partition` in `clojure.core`: for example `(partition n coll)` and `(partition n step pad coll)`. `step` defaults to `n`.

**Comment by hastebrot** [comment]
For educational purposes I implemented some variants for fixed and sliding windows [1]. I know there is already a prototype implemented (refered in the KEEP document) with test suite which even uses a `RingBuffer`.

- `partition()` implements a function for fixed and sliding windows suited for lists (not sequences) using indices/ranges.
- `partition1()` and `partition2()` implement such functions suited for lists using recursion and tail recursion (very inefficient).
- `batch()` and `slide()` implement such functions suited for lists and sequences (similar to the prototype implementation).

[1] https://gist.github.com/hastebrot/9b0532d9b9317f13d93aeb79c24125b4

A sliding window for pairs with an offset of one, can be implemented ad hoc with this simple variant which uses `zip()`:

~~~kotlin
fun main(args: Array<String>) {
    val items = sequenceOf(1, 2, 3, 4, 5)
    val windowItems = items.zip(items.drop(1))
    println(windowItems.toList()) // [(1, 2), (2, 3), (3, 4), (4, 5)]
}
~~~

**Comment by ilya-g** [comment]
Regarding to alternatives section: "copy-paste it everywhere" is not a meaningful alternative. I mean we can assume such an alternative for every standard library function proposal.

**Comment by ilya-g** [comment]
`slidingWindow2` is quite useful, usually it comes in form of something like `pairwise` or so. I think it deserves its own proposal.

On the opposite `slidingWindow3` seems to be a specialization of general `slidingWindow` for which I don't see compelling use cases.

**Comment by hastebrot** [comment]
@ilya-g `slidingWindow3` can be used to implement the Visvalingam-Whyatt polyline simplification algorithm [1]. In the algorithm in a list of points we use three consecutive points to build a triangle.

[1] https://bost.ocks.org/mike/simplify/

**Comment by ilya-g** [comment]
relates to `dropTrailing`: http://stackoverflow.com/questions/7958712/inconsistent-behaviour-for-xs-slidingn-if-n-is-less-than-size

**Comment by ilya-g** [comment]
Use case for `slidingWindow(size, step=1)`: https://github.com/jedrz/cdf/blob/318668f7e2304ebb1b08b9c6db15c7c158b38f5a/src/main/scala/cdf/matcher/ngrams/NGramsEvaluator.scala

**Comment by ilya-g** [comment]
Similar API: Groovy's collate 
http://mrhaki.blogspot.ru/2012/04/groovy-goodness-collate-list-into-sub.html
http://docs.groovy-lang.org/latest/html/groovy-jdk/java/lang/Iterable.html#collate(int)

**Comment by Groostav** [comment]
I feel its worth mentioning, we work with 2-dimensonal matricies through JNA from native code, meaning we get a lot of linearized 2D things --that is to say, a 2D grid converted into a 1D array by "stacking" rows one-after-the-other in an array. 

The standard strategy to convert them is

```kotlin
val columnCount: Int = ...
val input: Array<Double> = someNativeFunction()
vla matrix: List<List<Double>> = input.groupByIndexed { elem, index -> index / columnCount }
```

note that `groupByIndexed` is _not_ part of the STD lib, and is a function we wrote --we also use sequences a lot as these matrices are often quite large.

**Comment by ilya-g** [comment]
@Groostav `groupBy` returns a map, it would be strange and contradictory to other `*Indexed` functions  if `groupByIndexed` returned a list of lists.

**Comment by Groostav** [comment]
quite right @ilya-g , thus

```kotlin
public inline fun <T, K> Iterable<T>.groupByIndexed(keySelector: (index: Int, T) -> K): Map<K, List<T>> 
```

which would also satisfy my goals, if requiring a bit of reshaping for the output value.

**Comment by ilya-g** [comment]
I've revised the proposal, now there are three sets of functions considered: `chunked` to partition in non-overlapped blocks, `window` to partition in possibly overlapping windows or windows with gaps, and `pairwise` as a special case of `window` of size 2 shifted by 1 element to partition to pairs of adjacent elements.

**Comment by ilya-g** [comment]
> `fun <T, K> Iterable<T>.groupByIndexed(keySelector: (index: Int, T) -> K): Map<K, List<T>>`
> `fun <T, K> Sequence<T>.groupByIndexed(keySelector: (index: Int, T) -> K): Map<K, List<T>>`

@Groostav it's an interesting idea, but it's out of scope of current proposal. You can start a new proposal instead.

**Comment by mfulton26** [comment]
`s/pairwise/windowedPairwise/`?

It might be just me but if I read a function name of "pairwise" I wouldn't know if it took each pair one by one (chunked pairwise) or if it would window it (size of 2, step 1) or if it would do all-pairs ([Cartesian product](https://en.wikipedia.org/wiki/Cartesian_product)). If the name were changed to "windowedPairwise" then it would be immediately clear, I think, that this is a special case of "windowed" without even needing to read the documentation.

Thoughts?

**Comment by voddan** [comment]
When I first read `fun chunked(` and `fun windowed(`, I thought I saw `chuckled` and `widowed` 0_o

I am not a native english speaker, but it seems to me that certain rhythm is missing from those two words, which is why they sound a little unnatural.

**Comment by mfulton26** [comment]
I also wonder: Should this include implementing `forEachChunk`, `forEachWindow`, and `forEachPair` extension functions to iterate over the desired combinations without actually creating intermediate lists?

**Comment by ilya-g** [comment]
@mfulton26 That's how the overloads taking a lambda work.

**Comment by mfulton26** [comment]
That's right. :man_facepalming:  Thank you @ilya-g.

**Comment by marstran** [comment]
I find these functions very useful. One question: I keep thinking that the `step` parameter of the `windowed` function should have a default value of 1. Is that something that has been considered?

**Comment by voddan** [comment]
Why `pairwise` is `windowed(2, 1)`, not `chunked(2)`? 

As @mfulton26 [mentioned](https://github.com/Kotlin/KEEP/issues/11#issuecomment-311461094), I would expect a different behaviour from this function judging by its name. 

Is there any evidence that one is needed more often than another?

**Comment by voddan** [comment]
I checked it on the internet, and it seems that "pairwise" is generally used to describe all possible combinations of elements, not subsequent pairs of elements. 

The only place where "pairwise" means the same as in the proposal is the Python library, possibly because they have a `combinations` function as well.

I like the @marstran [suggestion](https://github.com/Kotlin/KEEP/issues/11#issuecomment-311978551) to add a default value of `1` to the `step` parameter of the `windowed`functions. If we implement it, then `pairwise` will be synonymous to `windowed(2)`. That would make `pairwise` unnecessary and save us a problem of coming up with a clearer name for this function.

**Comment by ilya-g** [comment]
> I checked it on the internet, and it seems that "pairwise" is generally used to describe all possible combinations of elements, not subsequent pairs of elements.

Could you share your findings? It would be especially useful if you could find an analogous function in some other framework, so we could compare the naming.

> If we implement it, then pairwise will be synonymous to windowed(2). That would make pairwise unnecessary and save us a problem of coming up with a clearer name for this function.

`pairwise` has some performance advantages over `windowed(2)`, so it cannot be replaced by the latter: it doesn't allocate a list to represent a window; also being a simplified case of `windowed` it has an implementation more suitable to be inlined. Therefore `pairwise` is inline, while `windowed` is not.

**Comment by ilya-g** [comment]
> I find these functions very useful. One question: I keep thinking that the step parameter of the windowed function should have a default value of 1. Is this something that has been considered?

@marstran Initially we had both `windowed` and `chunked` as a single method with `step` defaulted to window size, but after splitting it in two methods it became possible to make `step` default to `1`.

The pros of that should be weighed carefully, as it would double the method count occupied by `windowed`.
In case if we add another default parameter to `windowed`, like `dropPartialWindows = true`, it would be possible also to make `step` optional.

Is `step = 1` so common that it is a reasonable default?

**Comment by voddan** [comment]
>@voddan, could you share your findings?

I [searched StackOverflow for "pairwise"](https://stackoverflow.com/search?q=pairwise) and looked through the questions and how they were using the word. Excluding questions about Python, the majority of questions use "pairwise" to describe each-to-each relationships. 

I wasn't able to find any major framework that would use a "pairwise" function except for the Python library, possibly because of the controversy. Sometimes they use it in documentation to describe a pairwise comparison (which is each-to-each).

The wikipedia [says](https://en.wikipedia.org/wiki/Pairwise) that `"pairwise" generally means "occurring in pairs" or "two at a time"`, but in math the term everybody uses is [pairwise comparison](https://en.wikipedia.org/wiki/Pairwise_comparison), which is drastically different. I believe programmers prefer to use the mathematical meaning of this word.

**Comment by marstran** [comment]
I looked at some other languages to see what they do with functions like these:

Scala has the `sliding` function which is equivalent to the `windowed` function. This function has an overload with `step = 1`.

Ruby has the `each_cons` function which is equivalent to `windowed` with `step = 1`. I couldn't find a function in Ruby where you can provide the `step` parameter, but you have `each_slice` which is equivalent to `chunked`.

F# has the `windowed` function which is equivalent to Kotlin's `windowed` with `step = 1`. It doesn't take a `step` parameter. By the way, F# also has the `pairwise` function which is equivalent to Kotlin's `pairwise`.

Looking at all these, I find `step = 1` to be a reasonable default.

**Comment by ilya-g** [comment]
Alternative naming options for `pairwise`: 
- `pairWithNext`, 
- `zipWithNext` 
- `zipWithSelf(offset = 1)` (this option with adjustable offset leads to less effective implementation)

**Comment by ilya-g** [comment]
I've renamed `pairwise` to `zipWithNext` and changed the parameters of `windowed`. 
Now `step` defaults to 1 and there is a new parameter `partialWindows`, which controls whether to keep or drop incomplete windows in the end, defaults to false.
https://github.com/JetBrains/kotlin/compare/7c2dd7c210b85bcebfc7ca79850e3abf5c18df94~2...7c2dd7c210b85bcebfc7ca79850e3abf5c18df94

**Comment by voddan** [comment]
Nice! One small question thou: why `chunked` keeps partial windows, while `windowed` defaults to not having them? Isn't it a bit inconsistent?
I have no personal preference on this one.

**Comment by ilya-g** [comment]
@voddan The use case of `chunked` is to process a large collection in several smaller chunks. Usually you don't want the last elements of a collection to be dropped just because they form a chunk of the smaller size than requested. 
On the other hand window processing cases usually require to have a fixed size window, so it seems reasonable to default to dropping partial windows.

**Comment by fab1an** [comment]
Will there be support for "folding" function that group sequences into windows? Like having a function that determines when a window starts and another when the window ends resulting in a lazy sequence of these grouped items?

**Comment by hho** [comment]
I don't understand why `zipWithNext` uses a `step` of 1 instead of 2?

It's great to have a special case for `Pair`s, but intuitively I would expect _not_ to get every element twice.
 (And according to the Kotlin Slack, where this came up multiple times now, most people seem to share that assumption)

**Comment by ilya-g** [comment]
@fab1an We considered that function before, see https://github.com/Kotlin/KEEP/blob/master/proposals/stdlib/window-sliding.md#future-advancements, 
but decided that it would have to generic API and would be not much in demand.

But let's imagine: if you had such function, how would you call it? Could you show some code examples with this function?

**Comment by cy6erGn0m** [comment]
Btw, I tried to design and implement such function multiple times and every time it resulted in too complex declarations and difficult to understand parameters

**Comment by dzharkov** [comment]
The declarations has been released in Kotlin 1.2, thus I'm closing the KEEP. Do not hesitate to open a [YouTrack issue](https://kotl.in/issue) for any additional suggestions.
